Design network layer

cjsaez opened the pull request:
Please, if you can, take a look at my comments in the other PR.

rlaguilar commented on app/src/main/java/com/example/pelis/networkLayer/ApiRestOperations.kt:
```diff
@@ -0,0 +1,68 @@
+ * @param EngineResponseType is the type of the response that it be obtained
+ *                           as a result for call a request through the selected engine.
+ */
+interface ApiRestOperations<
```
Think of a better name for this type. `ApiRestOperations` doesn't really communicate anything about the purpose of the class. When you have a name ending in `s`, or in other words, the name is a plural, that usually means that you don't really have a clear purpose for that type, it basically does different things that aren't related to each other, but for some reason you decided to put them together in the same type.

rlaguilar commented on app/src/main/java/com/example/pelis/networkLayer/ApiRestOperations.kt:
```diff
@@ -0,0 +1,68 @@
+ *                           as a result for call a request through the selected engine.
+ */
+interface ApiRestOperations<
+        EngineRequestType : Any,
+        EngineResponseType : Any
```
You should get rid of these generic types because they add complexity to the class and don't help with the purpose you have (i.e. _to abstract the implementation for a selected networking engine_).

Let's say you have two engines A and B. Then you would have two sets of classes:

```kotlin
class AApiRestOperations: ApiRestOperations<AEngineRequest, AEngineResponse> {
    ...
}

class BApiRestOperations: ApiRestOperations<BEngineRequest, BEngineResponse> {
    ...
}
```

With your design, an instance of type `AApiRestOperations` is not interchangeable with an instance of type `BApiRestOperations`, so having the generic interface doesn't serve any purpose.

rlaguilar commented on app/src/main/java/com/example/pelis/networkLayer/ApiRestOperations.kt:
```diff
@@ -0,0 +1,68 @@
+     * @param MyResponseType is the type of the model that want to be obtained.
+     *
+     * @param[eResponse] is a response object that it be obtained by a engine request called.
+     * @param[type] is the common superinterface for all types in the Java.
```
This comment is wrong, it's describing the type `Type`  instead of describing the purpose of the parameter `type`.  It isn't clear to my why you also need that `type` parameter, so it would be great if you can add a detailed explanation of it in the comment :-)

rlaguilar commented on app/src/main/java/com/example/pelis/networkLayer/ApiRestOperations.kt:
```diff
@@ -0,0 +1,68 @@
+     * @param[type] is the common superinterface for all types in the Java.
+     * @return a specified model as a result of mapping the engine response.
+     */
+    suspend fun <MyResponseType : Any> parseEngineResponseType(
+        eResponse: EngineResponseType,
+        type: Type
+    ): MyResponseType
```
If you look at the framework, type names don't have the suffix `Type`. For example, you have `String`, not `StringType`. The same should also apply to your own types, it's gonna improve the readability of your code a lot. This example would look something like this:

```kotlin
suspend fun <MyResponse : Any> parseEngineResponse(
    response: EngineResponse,
    type: Type
): MyResponse
```

Notice that I'm using the name `response` instead of `eResponse`. When you name a variable, try to use a name that reflects the purpose of that variable, not its type. 

Also, don't use the prefix `My` for the types, use a name that have a meaning in the context you are using it. This part is hard, but you need to work hard trying to find good names for your types/variable/functions. The article I sent you the first day should have some tips to do that.

rlaguilar commented on app/src/main/java/com/example/pelis/networkLayer/NetworkEnginePerformRequest.kt:
```diff
@@ -0,0 +1,27 @@
+ * An interface to abstract the requesting resources to the network implementation
+ * for a selected networking engine.
+ */
+interface NetworkEnginePerformRequest {
```
Types represent entities, not actions. This means that type names should be nouns instead of verbs. In this case, `NetworkEnginePerformRequest` is a verb, `NetworkEngineRequestPerformer` would be a noun.

rlaguilar commented on app/src/main/java/com/example/pelis/networkLayer/Parser.kt:
```diff
@@ -0,0 +1,25 @@
+
+import java.lang.reflect.Type
+
+/**
+ * An interface to abstract the parsing of strings to the specified parser engine.
+ */
+interface Parser {
```
This comment doesn't make a lot of sense to me. In the whole interface there is nothing that point to the engine you mention in the comment.

rlaguilar commented on app/src/main/java/com/example/pelis/networkLayer/TypeReference.kt:
```diff
@@ -0,0 +1,22 @@
+import java.lang.reflect.ParameterizedType
+import java.lang.reflect.Type
+
+/**
+ * Represents a generic type [T]. Kotlin doesn't yet provide a way to
+ * represent generic types, so this class does. Forces clients to create a
+ * subclass of this class which enables retrieval the type information even at
+ * runtime.
+ *
+ * For example, to create a type literal for List<String>, you can
+ * create an empty anonymous inner class:
+ *
+ * val list: TypeToken<List<String>> = TypeToken<List<String>>() {};}
+ **/
+abstract class TypeReference<T> : Comparable<TypeReference<T>> {
+    val type: Type =
+        (javaClass.genericSuperclass as ParameterizedType).actualTypeArguments[0]
+
+    override fun compareTo(other: TypeReference<T>) = 0
+}
```
Could you please provide clearer documentation for this type? It's quite cryptic and it's not clear what problem is solves. You say:

> Kotlin doesn't yet provide a way to represent generic types, so this class does.

That's not true, Kotlin does provide support for generic types, in fact, `TypeReference<T>` is a generic type.

cjsaez replied:
Evidently Kotlin support genericity, the problem is obtain the Class<T> at runtime when T is generic (ex: T instaceOf List<Person>, the compile recognize 'List' but not 'Person'). The parser engines like Gson has a kind of class like this (TypeToken in Gson)

cjsaez replied:
...above comment, in the ex. it is List<Person>

cjsaez replied:
the type is list of Persone, I do not know how put code in these comments

rlaguilar replied:
The comment is still wrong, please fix it :-) 

> I do not know how put code in these comments

You can always google it. Here's [a guide](https://docs.github.com/en/get-started/writing-on-github) to do that and other things.

rlaguilar replied:
I think it would help if you explain how you are planning to use this type with a sample code. It seems you are planning to use it on the parser, could you please show a code snippet of how that would look like?

cjsaez commented:
> I think it would help if you explain how you are planning to use this type with a sample code. It seems you are planning to use it on the parser, could you please show a code snippet of how that would look like?

The comment was fixed, I think that with it, the comments of the class and the comments of `performRequest` and `parse` functions you can to make a better idea of why this class is needed; if not let me know.

rlaguilar commented:
I put this comment above:
> > Kotlin doesn't yet provide a way to represent generic types, so this class does.
>
> That's not true, Kotlin does provide support for generic types, in fact, TypeReference<T> is a generic type.

That comment is still in the code, I asked you to fix it:

> The comment is still wrong, please fix it :-)

The comment is still in the code. 

The comment you fixed was wrong, but had nothing to do with anything I mentioned before. Please, try to pay attention to the conversation so that we can move forward instead of spending time talking about the same thing over and over again. 


> The comment was fixed, I think that with it, the comments of the class and the comments of performRequest and parse functions you can to make a better idea of why this class is needed

From the time I asked you to provide a code snippet to the time you put this reply, the only thing that changed was the code that you fixed, which was unrelated to everything we've been talking about. How will that in any way help me understand what do you need that class for? I already said what would help me to gain understanding: 

> It seems you are planning to use it on the parser, could you please show a code snippet of how that would look like?

This is the function declared in the `Parser` interface:
```kotlin
suspend fun <T : Any> parse(textToParse: String, typeOfT: Type): T
```

A code snippet would be an example implementation of that function, so that I could see why you need the parameter `typeOfT: Type`.

cjsaez commented:
I think there's a misunderstanding, the comment in the code that I fixed is not related with your comment like you said, I was refer to my comment: 
> Evidently Kotlin support genericity, the problem is obtain the `Class<T>` at runtime when T is generic (ex: `T instaceOf List<Person>`, the compile recognize `List` but not `Person`). The parser engines like Gson has a kind of class like this (`TypeToken` in Gson)   
   
I putted this comment and two more because I didn't know how to put code in this comments, followed them you put: 
> The comment is still wrong, please fix it :-)

and in that point I hadn't touch the class comment, for that, I think that you was refer to the PR comment that I put here because you wanted to understand it. I fixed this comment that show the problem resolved by this class and deleted the unnecessary other two.

rlaguilar commented on app/src/main/java/com/example/pelis/networkLayer/NetworkEngineRequestPerformer.kt:
```diff
@@ -0,0 +1,27 @@
+     *
+     * @return a specified model as a result of mapping the engine response to [typeOfT] encapsulate in a Kotlin.Result instance.
+     */
+    suspend fun <T> performRequest(request: Request, typeOfT: Type): Result<T>
```
By the name and comments in the class, it seems to me that the responsibility of this class is to perform network requests. You could have different implementations, each one performing the requests in its own way. 

However, when you make this method generic, you are saying that this class will also take care of parsing the request response, which is outside the scope of the class. You could either make this function return a `Result<String>`, or create a richer type adding more information to the response:

```kotlin
data class Response(val statusCode: Int, val body: String)
```

and then return a `Result<Response>` object. The advantage of creating this `Response` type is that you could have a validation mechanism shared across all the `RequestPerformer` implementations.

rlaguilar commented on app/src/main/java/com/example/pelis/networkLayer/KtorRequestPerformerImpl.kt:
```diff
@@ -0,0 +1,69 @@
+        when (request.requestMethod) {
+            RequestMethod.GET -> Get
+            RequestMethod.POST -> Post
+            else -> Put
```
Try to handle enum cases specifically as much as you can. In this case you have two options, both working the same at the moment:

1. `else -> Put`
2. `RequestMethod.PUT -> Put`

Now, imagine that in the future you add `RequestMethod.DELETE`. If you are using the first option, the project will continue compiling just fine, and the logic inside this function will be wrong because it will assign `Put` to `DELETE`. However, if you use the second option, you will get a compiler error in all the places that you are doing a switch over the enum, which will help you to update the code to handle the new case.

rlaguilar reviewed (changes requested):
Thanks for updating the comment and providing more information. It helped to understand the purpose of the type. 

The project isn't compiling in this branch, make sure it does.
